src/modules/adaptors/Balancer/BalancerPoolAdaptor.sol

```diff
@@ -0,0 +1,295 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.16;
+
+import { BaseAdaptor, ERC20, SafeTransferLib, Cellar, SwapRouter, Registry } from "src/modules/adaptors/BaseAdaptor.sol";
+import { IBalancerQueries } from "src/interfaces/external/Balancer/IBalancerQueries.sol";
+import { IVault } from "src/interfaces/external/Balancer/IVault.sol";
+import { IBalancerRelayer } from "src/interfaces/external/Balancer/solidity-utils/IBalancerRelayer.sol";
+import { IStakingLiquidityGauge } from "src/interfaces/external/Balancer/IStakingLiquidityGauge.sol";
+import { IBalancerRelayer } from "src/interfaces/external/Balancer/IBalancerRelayer.sol";
+import { ILiquidityGaugeFactory } from "src/interfaces/external/Balancer/ILiquidityGaugeFactory.sol";
+import { ILiquidityGaugev3Custom } from "src/interfaces/external/Balancer/ILiquidityGaugev3Custom.sol";
+
+/**
+ * @title Balancer Pool Adaptor
+ * @notice Allows Cellars to interact with Weighted, Stable, and Linear Balancer Pools (BPs).
+ * @author 0xEinCodes and CrispyMangoes
+ * TODO: This contract is still a WIP, Still need to go through TODOs and resolve relevant github issues
+ * TODO: Add event emissions where necessary
+ */
+contract BalancerPoolAdaptor is BaseAdaptor {
+    using SafeTransferLib for ERC20;
+
+    //==================== Adaptor Data Specification ====================
+    // adaptorData = abi.encode(ERC20 _bpt, address _poolId, JoinPoolRequest _request)
```

Comment:
I think you can get away with just having the bpt address as the adaptor data

Reply:
Ahhh I see. I was getting `adaptorCalls` mixed up with `adaptorData` that is added to the `Registry`.

Is `assetOf()` supposed to only return a single `ERC20` always? If so, what `ERC20` will it be for entering `BoostedPools` or other pools consisting of nested Balancer Pools? IIRC the `_accounting` from `Cellar.sol` will go through all `debt` and `credit` positions and use the `_assetOf()` to get the `ERC20` to use for pricing. This would be custom to the respective adaptor, `BalancerPoolAdaptor.sol` in this case.

I have to double check how the `Extension` design is working, but does it break down yield bearing tokens to base-constituents? AKA can it receive a boosted pool bpt and break it down to 1 ERC20 token (like DAI from bb-a-USD)?

Reply:
For now I'll make it just the bpt address, but I think we need to discuss what I was talking about above more / your expertise would be great :)

Reply:
The extension does break down yield bearing tokens into their underlyings in order to find a USD value of BPTs